### DISCUSSION TITLE
About ua-parser-js

### DISCUSSION THREAD

**Discussion by LeviticusMB** [pnpm/pnpm#3913] [Q&A] [answered]
So yesterday evening, I decided to bump the deps in a project of mine. Checked with `pnpm outdated -r` first, then `pnpm upgrade -r`.

After this, I had two versions of `ua-parser-js` in `node_modules/.pnpm`: `ua-parser-js@0.7.28` and `ua-parser-js@0.7.30`. But since `pnpm outdated -r` also listed version 1.0.1 as the latest version, I decided to check out what was new and quickly discovered faisalman/ua-parser-js#536.

Had I run the upgrade command just a few hours earlier, well ... then what? What would have happened?

1. How is `pnpm` protecting me from a game over scenario when running `pnpm upgrade`?
2. What can I do myself to ensure nothing bad happens in this case?

**Comment by zkochan** [comment]
I am not sure what we can do on our side. Maybe allow new versions with a delay? So only upgrade to a version if it was released 24 hours ago?

**Reply by LeviticusMB** [reply]
A configurable grace period actually sounds like a *great* idea. That would allow all kinds of static analysis and auditing to packages before they are actually used.

I also quickly checked a few of my project dependencies and only 5 packages out of 2100 used some kind of install script. 2 out of 1700 in another project. How about a whilelist for scripts? If a package all of a sudden began to use an install script, it could be audited before being allowed to run.

Sure, bad code in a lib would *eventually* execute anyway, but chances are it will be in a container, locked down prod/staging environment or even in the browser and not my laptop — and not instantly.

**Reply by zkochan** [reply]
We can do something like executing the scripts of "trusted" packages only.

**Reply by LeviticusMB** [reply]
Another thing that came to mind: Would it be possible to flag those packages, in `pnpm audit` perhaps, where not all with publishing rights have 2FA enabled? Is that information available somewhere? That could be helpful when deciding what dependencies to use and trust in a project.

I understand that if there was an easy solution to the problem, there would be no problem. But every little thing helps.

**Reply by zkochan** [reply]
I don't think the npm registry gives us that information. I hope not, it would give attackers advantage.

**Reply by zkochan** [reply]
Kind of a related discussion: https://github.com/pnpm/pnpm/issues/3881

I think it would make sense to more prominently display the publisher's identity.

**Reply by LeviticusMB** [reply]
Just wanted to say that I'm really happy that `onlyBuiltDependencies` now exists, and that it even defaults to nothing in v10! Awesome.

I still think that the grace period is a good idea though. And the 2FA warning.

**Reply by zkochan** [reply]
Good, looks like there is a loud angry minority that is not happy with the change. Had to lock a few discussions already.

**Reply by LeviticusMB** [reply, marked as answer]
And now #9921! 💯
